Warn on duplicate cases in switch statements

Rebraws opened the pull request:
Now quick-lint-js gives a warning when a switch statement has duplicated cases.

closes  https://github.com/quick-lint/quick-lint-js/issues/744

strager commented on src/quick-lint-js/fe/diagnostic-types.h:
```diff
@@ -464,6 +464,13 @@
+      diag_duplicated_cases_in_switch_statement, "E0347",                       \
+      diagnostic_severity::warning,                                             \
+      { source_code_span duplicated_switch_case; },                             \
+      MESSAGE(QLJS_TRANSLATABLE("duplicated case clause in switch statement"),  \
+              duplicated_switch_case))                                          \
```
Important: We should point to the original `case` too:

> \<location of second case>: duplicated case clause in switch statement
> \<location of first case>: this case will run instead

strager commented on src/quick-lint-js/fe/parse-statement.cpp:
```diff
@@ -1851,10 +1853,19 @@ void parser::parse_and_visit_switch(parse_visitor_base &v) {
                });
+
+        source_code_span expression_case_span = ast->span();
+        if (repeated_cases.contains(expression_case_span.string_view())) {
+          this->diag_reporter_->report(
+              diag_duplicated_cases_in_switch_statement{
+                  .duplicated_switch_case = expression_case_span});
+        }
+        repeated_cases.insert(expression_case_span.string_view());
```
Perf: We shouldn't do two lookups. Instead, use `insert` and see if it failed:
```suggestion
        auto [inserted, _it] = repeated_cases.insert(expression_case_span.string_view());
        if (!inserted) {
          this->diag_reporter_->report(
              diag_duplicated_cases_in_switch_statement{
                  .duplicated_switch_case = expression_case_span});
        }
```

strager commented on test/test-parse-statement.cpp:
```diff
@@ -949,6 +949,42 @@ TEST_F(test_parse_statement, switch_case_without_expression) {
+    test_parser p(u8"switch (cond) {case x: case y: case y:}"_sv,
+                  capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.visits, ElementsAre("visit_variable_use",       // cond
+                                      "visit_enter_block_scope",  //
+                                      "visit_variable_use",       // x
+                                      "visit_variable_use",       // y
+                                      "visit_variable_use",       // y
+                                      "visit_exit_block_scope"));
+    EXPECT_THAT(p.errors,
+                ElementsAre(DIAG_TYPE_OFFSETS(
+                    p.code, diag_duplicated_cases_in_switch_statement,
+                    duplicated_switch_case,
+                    strlen(u8"switch (cond) {case x: case y: case "), u8"y")));
+  }
+  {
+    test_parser p(
+        u8"switch (cond) {case MyEnum.A: break; case MyEnum.A: break;}"_sv,
+        capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.visits, ElementsAre("visit_variable_use",       // cond
+                                      "visit_enter_block_scope",  //
+                                      "visit_variable_use",       // MyEnum.A
+                                      "visit_variable_use",       // MyEnum.A
+                                      "visit_exit_block_scope"));
```
Nit: No need to assert visits, since we're just checking for duplicated cases.
```suggestion
    EXPECT_THAT(p.errors,
                ElementsAre(DIAG_TYPE_OFFSETS(
                    p.code, diag_duplicated_cases_in_switch_statement,
                    duplicated_switch_case,
                    strlen(u8"switch (cond) {case x: case y: case "), u8"y")));
  }
  {
    test_parser p(
        u8"switch (cond) {case MyEnum.A: break; case MyEnum.A: break;}"_sv,
        capture_diags);
    p.parse_and_visit_statement();
```

strager commented on src/quick-lint-js/fe/parse-statement.cpp:
```diff
@@ -1834,6 +1835,7 @@ void parser::parse_and_visit_switch(parse_visitor_base &v) {
 
   bool keep_going = true;
   bool is_before_first_switch_case = true;
+  hash_set<string8_view> repeated_cases;
```
Nit: I think a better name is `cases` or `case_expressions`. `repeated_cases` sounds like it contains *only* the duplicates, but it actually contains *every* `case`.

strager reviewed (approved):
Looks nice. Much simpler than I thought it would be.

Rebraws commented:
I think I've addressed all the comments. I had to replace `hash_set<string8_view>` with `hash_map<string8_view, source_code_span>` to also keep track of all the `source_code_span` to point to the original `case` when giving the warning.

Please let me know if anything else needs to be changed.

strager commented:
>  I had to replace `hash_set<string8_view>` with `hash_map<string8_view, source_code_span>`

I think this is unnecessary. source_code_span can be converted into string8_view, and string8_view can be converted back to source_code_span. Just `hash_set<string8_view>` should be fine.

strager commented on test/test-parse-statement.cpp:
```diff
@@ -949,6 +949,31 @@ TEST_F(test_parse_statement, switch_case_without_expression) {
+                  capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.errors,
+                ElementsAre(DIAG_TYPE_OFFSETS(
+                    p.code, diag_duplicated_cases_in_switch_statement,
+                    duplicated_switch_case,
+                    strlen(u8"switch (cond) {case x: case y: case "), u8"y")));
```
Please test `first_switch_case` too.

strager commented on src/quick-lint-js/fe/parse-statement.cpp:
```diff
@@ -1851,10 +1853,21 @@ void parser::parse_and_visit_switch(parse_visitor_base &v) {
+          this->diag_reporter_->report(
+              diag_duplicated_cases_in_switch_statement{
+                  .first_switch_case =
+                      source_code_span(_it->begin(), _it->end()),
```
`begin` and `end` return iterators, but `source_code_span` expects pointers. This is causing build failures with MSVC. You should use `_it->data()` and `_it->data() + _it->size()` instead.

Rebraws replied:
Thank you, I didn't realized that the constructor expected pointers, now I was just looking to the `source_code_span` class and saw that there's a `source_code_span::unit` member function, since we are using this span only to point to the original `case` do you think it would be better to use that function like?
`.first_switch_case = source_code_span::unit(_it->data())`

Or should I just use the constructor

strager replied:
No, you don't want to use `unit`. `unit` is for diagnostics which point between two characters. Use the two-pointer constructor.

Rebraws commented:
Fixed the CI error and added the tests for `first_switch_case` let me know if there's anything else to improve :)

strager commented on test/test-parse-statement.cpp:
```diff
@@ -949,6 +949,33 @@ TEST_F(test_parse_statement, switch_case_without_expression) {
+    EXPECT_THAT(p.errors,
+                ElementsAre(DIAG_TYPE_2_OFFSETS(
+                    p.code, diag_duplicated_cases_in_switch_statement,
+                    first_switch_case, strlen(u8"switch (cond) {case x: case "),
+                    u8"y", duplicated_switch_case,
```
Nit: Add a blank comment to improve code formatting:
```suggestion
                    first_switch_case, strlen(u8"switch (cond) {case x: case "),
                    u8"y", //
                    duplicated_switch_case,
```

Rebraws replied:
Fixed it  in commit https://github.com/quick-lint/quick-lint-js/pull/866/commits/0172714a51675810905d3cd0f84e72daef5d1078 and I also did a rebase to fix the merge conflicts

strager commented:
Beautiful!
